gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/matching/single/cypher/operators/filter/functions/FilterAndProjectTriple.java

```diff
@@ -0,0 +1,180 @@
+   * @param targetVariable target variable
+   * @param predicates filter predicates
+   * @param projectionPropertyKeys property keys used for projection
+   * @param vertexMatchStrategy vertex match strategy
+   */
+  public FilterAndProjectTriple(String sourceVariable, String edgeVariable, String targetVariable,
+    CNF predicates, Map<String, List<String>> projectionPropertyKeys,
+    MatchStrategy vertexMatchStrategy) {
+
+    this.predicates = predicates;
+    this.sourceVariable = sourceVariable;
+    this.targetVariable = targetVariable;
+    this.vertexMatchStrategy = vertexMatchStrategy;
+
+    this.sourceProjectionPropertyKeys =
+      projectionPropertyKeys.getOrDefault(sourceVariable, new ArrayList<>());
+    this.edgeProjectionPropertyKeys =
+      projectionPropertyKeys.getOrDefault(edgeVariable, new ArrayList<>());
+    this.targetProjectionPropertyKeys =
+      projectionPropertyKeys.getOrDefault(targetVariable, new ArrayList<>());
+
+    filterMetaData = createFilterMetaData(predicates, sourceVariable, edgeVariable, targetVariable);
+    sourceFilterPropertyKeys = filterMetaData.getPropertyKeys(sourceVariable);
+    edgeFilterPropertyKeys = filterMetaData.getPropertyKeys(edgeVariable);
+    targetFilterPropertyKeys = filterMetaData.getPropertyKeys(targetVariable);
+
+  }
+
+  @Override
+  public void flatMap(Triple triple, Collector<Embedding> out) throws Exception {
+
+    if (sourceVariable.equals(targetVariable)) {
+      // Check if source and target are expected to be the same
+      if (!triple.getSourceVertex().getId().equals(triple.getTargetVertex().getId())) {
+        return;
+      }
+    } else if (vertexMatchStrategy.equals(MatchStrategy.ISOMORPHISM) &&
+      triple.getSourceVertex().getId().equals(triple.getTargetVertex().getId())) {
+      return;
+    }
+
+    if (filter(triple)) {
+      out.collect(
+        EmbeddingFactory.fromTriple(
+          triple,
+          sourceProjectionPropertyKeys, edgeProjectionPropertyKeys, targetProjectionPropertyKeys,
+          sourceVariable, targetVariable
+        )
+      );
+    }
+  }
+
+  /**
+   * Checks if the the triple holds for the predicate
+   * @param triple triple to be filtered
+   * @return True if the triple holds for the predicate
+   */
+  private boolean filter(Triple triple) {
+    return predicates.evaluate(
+      EmbeddingFactory.fromTriple(triple,
+        sourceFilterPropertyKeys,  edgeFilterPropertyKeys, targetFilterPropertyKeys,
+        sourceVariable, targetVariable
+      ),
+      filterMetaData
+    );
+  }
+
+  /**
+   * Creates the {@code EmbeddingMetaData} of the embedding used for filtering
+   * @param predicates filter predicates
+   * @param sourceVariable source variable
+   * @param edgeVariable edge variable
+   * @param targetVariable target variable
+   * @return filter embedding meta data
+   */
+  private static EmbeddingMetaData createFilterMetaData(CNF predicates, String sourceVariable,
```

Comment:
But for any other physical operator the `EmbeddingMetaData` is updated in the corresponding Node class, isn't it?